binstat/binstat.go

```diff
@@ -0,0 +1,546 @@
+			}
+		}
+	}
+	global.key2index = make(map[string]int)
+
+	appendInternalKey := func(keyIotaIndex int, name string) {
+		if keyIotaIndex != len(global.keysArray) {
+			panic(fmt.Sprintf("ERROR: BINSTAT: INTERNAL: %s", name))
+		}
+		global.keysArray = append(global.keysArray, name)
+		global.keysEgLoc = append(global.keysEgLoc, "")
+		global.frequency = append(global.frequency, 0)
+		global.accumMono = append(global.accumMono, 0)
+		global.frequencyShadow = append(global.frequencyShadow, 0)
+		global.accumMonoShadow = append(global.accumMonoShadow, 0)
+	}
+	appendInternalKey(internalSec, "/internal/second")
+	appendInternalKey(internalEnter, "/internal/binstat.enter")
+	appendInternalKey(internalPoint, "/internal/binstat.point")
+	appendInternalKey(internalx_2_y, "/internal/binstat.x_2_y")
+	appendInternalKey(internalDebug, "/internal/binstat.debug")
+	appendInternalKey(internalGC, "/internal/GCStats")
+
+	global.index = len(global.keysArray)
+	global.indexInternalMax = len(global.keysArray)
+	go tick(100 * time.Millisecond) // every 0.1 seconds
+
+	if global.verbose {
+		elapsedThisProc := time.Duration(runtimeNanoAsTimeDuration() - global.startTimeMono).Seconds()
+		globalLog.Debug().Msgf("%f %d=pid %d=tid init() // .enable=%t .verbose=%t .dmpPath=%s .dmpName=%s .cutPath=%s .lenWhat=%s",
+			elapsedThisProc, os.Getpid(), int64(C.gettid()), global.enable, global.verbose, global.dmpPath, global.dmpName, global.cutPath, global.lenWhat)
+	}
+
+	t2 := runtimeNanoAsTimeDuration()
+	if t2 <= t1 {
+		panic(fmt.Sprintf("ERROR: BINSTAT: INTERNAL: t1=%d but t2=%d\n", t1, t2))
+	}
+}
+
+func Fin() {
+	dump()
+	// todo: consider closing down more somehow?
+}
+
+func enterGeneric(what string, callerParams string, callerTime bool, callerSize int64, callerSizeWhen int, verbose bool) *BinStat {
+	if !global.enable {
+		return nil
+	}
+
+	t := runtimeNanoAsTimeDuration()
+
+	funcName := ""
+	fileLine := 0
+	if global.verbose {
+		// todo: is there a way to speed up runtime.Caller() and/or runtime.FuncForPC() somehow? cache pc value or something like that?
+		pc, _, lineNum, _ := runtime.Caller(2) // 2 assumes private binStat.newGeneric() called indirectly via public stub function; please see eof
+		fn := runtime.FuncForPC(pc)
+		funcName = fn.Name()
+		funcName = strings.ReplaceAll(funcName, global.cutPath, "")
+		fileLine = lineNum
+	}
+
+	whatLen := len(what)
+	if (what[0:1] == "~") && (len(what) >= 3) {
+		// come here if what is "~<default len><what>", meaning that BINSTAT_LEN_WHAT may override <default len>
+		whatLenDefault := atoi(what[1:2])
+		whatLen = whatLenDefault + 2
+		if whatLenOverride, keyExists := global.what2len[what[2:2+whatLenDefault]]; keyExists {
+			whatLen = whatLenOverride + 2
+		}
+		if whatLen > len(what) {
+			whatLen = len(what)
+		}
+		//debug fmt.Printf("debug: detected ~%d%s in %s; using ~%d..\n", whatLenDefault, what[2:2 + whatLenDefault], what, whatLen)
```

Comment:
@turbolent thanks for the comment. Generally I like to keep things like this around for future debugging by the author or others. Plus it gives extra context for code comprehension. But I deleted it because there is only one line like that so far, and I don't want to draw out this PR :-)